source/Target/SwiftLanguageRuntime.cpp

```diff
@@ -2001,57 +2001,28 @@ bool SwiftLanguageRuntime::GetDynamicTypeAndAddress_Promise(
-    // Error has a special, NSError-compatible layout.
     SwiftASTContext *swift_ast_ctx =
         llvm::dyn_cast_or_null<SwiftASTContext>(protocol_type.GetTypeSystem());
-    if (swift_ast_ctx && swift_ast_ctx->IsErrorType(protocol_type)) {
-      Status error;
-      lldb::addr_t archetype_ptr_value = in_value.GetValueAsUnsigned(0);
-      lldb::addr_t metadata_ptr =
-          m_process->ReadPointerFromMemory(archetype_ptr_value, error);
-      MetadataPromiseSP promise_sp(GetMetadataPromise(metadata_ptr, in_value));
-      if (!promise_sp)
-        return false;
-      CompilerType dynamic_type(promise_sp->FulfillTypePromise());
-      if (dynamic_type.IsValid()) {
-        class_type_or_name.SetCompilerType(dynamic_type);
-        address.SetLoadAddress(archetype_ptr_value, &m_process->GetTarget());
-        return true;
-      }
-    } else {
-      Status error;
-      lldb::addr_t ptr_to_instance_type = in_value.GetValueAsUnsigned(0) +
-                                          (3 * m_process->GetAddressByteSize());
-      lldb::addr_t metadata_of_impl_addr =
-          m_process->ReadPointerFromMemory(ptr_to_instance_type, error);
-      if (error.Fail() || metadata_of_impl_addr == 0 ||
-          metadata_of_impl_addr == LLDB_INVALID_ADDRESS)
-        return false;
-      MetadataPromiseSP promise_of_impl_sp(
-          GetMetadataPromise(metadata_of_impl_addr, in_value));
-      if (GetDynamicTypeAndAddress_Promise(in_value, promise_of_impl_sp,
-                                           use_dynamic, class_type_or_name,
-                                           address)) {
-        lldb::addr_t load_addr = in_value.GetValueAsUnsigned(0);
-        if (promise_of_impl_sp->FulfillKindPromise() &&
-            promise_of_impl_sp->FulfillKindPromise().getValue() ==
-                swift::MetadataKind::Class) {
-          load_addr = m_process->ReadPointerFromMemory(load_addr, error);
-          if (error.Fail() || load_addr == 0 ||
-              load_addr == LLDB_INVALID_ADDRESS)
-            return false;
-        } else if (promise_of_impl_sp->FulfillKindPromise() &&
-                   (promise_of_impl_sp->FulfillKindPromise().getValue() ==
-                        swift::MetadataKind::Enum ||
-                    promise_of_impl_sp->FulfillKindPromise().getValue() ==
-                        swift::MetadataKind::Struct)) {
-        } else
-          lldbassert(false && "class, enum and struct are the only protocol "
-                              "implementor types I know about");
-        address.SetLoadAddress(load_addr, &m_process->GetTarget());
-        return true;
-      }
-    }
+    lldb::addr_t existential_address = in_value.GetPointerValue();
+    if (!existential_address || existential_address == LLDB_INVALID_ADDRESS)
+      return false;
+    auto &target = m_process->GetTarget();
+    assert(IsScratchContextLocked(target) &&
+           "Swift scratch context not locked ahead");
```

Comment:
@adrian-prantl Adrian, I'll appreciate if you can take a look and let me know whether the scratchASTContext acquisition is correct (I think so, but another pair of eyes is always welcome).

Reply:
LGTM.